File: lib/features/clan/models/clan_join_leave.dart
import 'package:clashkingapp/core/constants/image_assets.dart';

class ClanJoinLeave {
  final String clanTag;
  final int available;
  final int uniquePlayers;
  final List<JoinLeaveEvent> joinLeaveList;

  ClanJoinLeave({
    required this.clanTag,
    required this.available,
    required this.uniquePlayers,
    required this.joinLeaveList,
  });

  factory ClanJoinLeave.fromJson(Map<String, dynamic> json) {
    return ClanJoinLeave(
      clanTag: json['clan_tag'] ?? "",
      available: (json['available'] as num?)?.toInt() ?? 0,
      uniquePlayers: (json['uniquePlayers'] as num?)?.toInt() ?? 0,
      joinLeaveList:
          (json['items'] as List<dynamic>? ??
                  json['join_leave_list'] as List<dynamic>? ??
                  const [])
              .map((e) => JoinLeaveEvent.fromJson(e))
              .toList(),
    );
  }

  factory ClanJoinLeave.empty() {
    return ClanJoinLeave(
      clanTag: "",
      available: 0,
      uniquePlayers: 0,
      joinLeaveList: [],
    );
  }

  ClanJoinLeave appendPage(ClanJoinLeave page) {
    final seen = <String>{};
    final combined = <JoinLeaveEvent>[];
    for (final event in [...joinLeaveList, ...page.joinLeaveList]) {
      final key =
          '${event.time.toUtc().toIso8601String()}|${event.type}|${event.tag}';
      if (seen.add(key)) combined.add(event);
    }
    return ClanJoinLeave(
      clanTag: clanTag.isEmpty ? page.clanTag : clanTag,
      available: page.available,
      uniquePlayers: page.uniquePlayers,
      joinLeaveList: combined,
    );
  }
}

class JoinLeaveEvent {
  final String type;
  final JoinLeaveClan? clan;
  final DateTime time;
  final String tag;
  final String name;
  final int th;

  JoinLeaveEvent({
    required this.type,
    required this.clan,
    required this.time,
    required this.tag,
    required this.name,
    required this.th,
  });

  factory JoinLeaveEvent.fromJson(Map<String, dynamic> json) {
    return JoinLeaveEvent(
      type: json['type'] ?? "",
      clan: json['clan'] is Map<String, dynamic>
          ? JoinLeaveClan.fromJson(json['clan'] as Map<String, dynamic>)
          : null,
      time: DateTime.tryParse(json['time'] ?? '') ?? DateTime(1970),
      tag: json['tag'] ?? "",
      name: json['name'] ?? "",
      th:
          (json['townHallLevel'] as num?)?.toInt() ??
          (json['th'] as num?)?.toInt() ??
          0,
    );
  }
}

class JoinLeaveClan {
  final String name;
  final String tag;
  final String badge;

  const JoinLeaveClan({
    required this.name,
    required this.tag,
    required this.badge,
  });

  factory JoinLeaveClan.fromJson(Map<String, dynamic> json) {
    final tag = json['tag'] as String? ?? '';
    return JoinLeaveClan(
      name: json['name'] as String? ?? '',
      tag: tag,
      badge: ImageAssets.clanBadgeForTag(tag),
    );
  }
}

File: expo/src/features/clan/models/clan-join-leave.ts
import { JoinLeaveClan } from './clan-core';
import { apiDateOrEpoch, int, record, records, string, type JsonRecord } from './parsing';

export class JoinLeaveEvent {
  constructor(
    readonly type: string,
    readonly clan: JoinLeaveClan | null,
    readonly time: Date,
    readonly tag: string,
    readonly name: string,
    readonly th: number,
  ) {}

  static fromJson(json: JsonRecord): JoinLeaveEvent {
    const clan = record(json.clan);
    return new JoinLeaveEvent(
      string(json.type),
      Object.keys(clan).length ? JoinLeaveClan.fromJson(clan) : null,
      apiDateOrEpoch(json.time),
      string(json.tag),
      string(json.name),
      int(json.townHallLevel, int(json.th)),
    );
  }
}

export class ClanJoinLeave {
  constructor(
    readonly clanTag: string,
    readonly available: number,
    readonly uniquePlayers: number,
    readonly joinLeaveList: readonly JoinLeaveEvent[],
  ) {}

  static fromJson(json: JsonRecord): ClanJoinLeave {
    // join_leave_list is a deliberate Flutter fallback for the prior response shape.
    const items = Array.isArray(json.items) ? json.items : json.join_leave_list;
    return new ClanJoinLeave(
      string(json.clan_tag),
      int(json.available),
      int(json.uniquePlayers),
      records(items).map(JoinLeaveEvent.fromJson),
    );
  }

  static empty(): ClanJoinLeave {
    return new ClanJoinLeave('', 0, 0, []);
  }

  appendPage(page: ClanJoinLeave): ClanJoinLeave {
    const seen = new Set<string>();
    const combined: JoinLeaveEvent[] = [];
    for (const event of [...this.joinLeaveList, ...page.joinLeaveList]) {
      const key = `${event.time.toISOString()}|${event.type}|${event.tag}`;
      if (!seen.has(key)) {
        seen.add(key);
        combined.push(event);
      }
    }
    return new ClanJoinLeave(
      this.clanTag || page.clanTag,
      page.available,
      page.uniquePlayers,
      combined,
    );
  }
}
